crates/simulations/src/cozy/agents/arbitrageur.rs

```diff
@@ -0,0 +1,232 @@
+use std::{borrow::Cow, sync::Arc};
+
+use rand::{rngs::StdRng, Rng};
+use simulate::{
+    address::Address,
+    agent::{agent_channel::AgentChannelSender, Agent},
+    state::State,
+    u256::{U256, *},
+    utils::decode_output,
+};
+
+use crate::cozy::{
+    decay_normalizer::normalize_constant_decay_price,
+    runner::{ProtocolContracts, SetContracts},
+    set_risk_model::SetRiskModel,
+    types::ReactionTime,
+    world::{CozyUpdate, CozyWorld},
+};
+
+#[derive(Debug, Clone)]
+pub struct ArbitrageurPreferences {
+    risk_model: SetRiskModel,
+    reaction_time: ReactionTime,
+}
+
+impl ArbitrageurPreferences {
+    pub fn new(risk_model: SetRiskModel, reaction_time: ReactionTime) -> Self {
+        Self {
+            risk_model,
+            reaction_time,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Arbitrageur {
+    _name: Cow<'static, str>,
+    address: Address,
+    protocol: Arc<ProtocolContracts>,
+    set: Arc<SetContracts>,
+    preferences: ArbitrageurPreferences,
+    rng: StdRng,
+}
+
+impl Arbitrageur {
+    pub fn new(
+        _name: Cow<'static, str>,
+        address: Address,
+        protocol: Arc<ProtocolContracts>,
+        set: Arc<SetContracts>,
+        preferences: ArbitrageurPreferences,
+        rng: StdRng,
+    ) -> Self {
+        Self {
+            _name,
+            address,
+            protocol,
+            set,
+            preferences,
+            rng,
+        }
+    }
+}
+
+impl Agent<CozyUpdate, CozyWorld> for Arbitrageur {
+    fn address(&self) -> Address {
+        self.address
+    }
+
+    fn activation_step(&mut self, state: &mut State<CozyUpdate, CozyWorld>) {
+        let router_approve_tx = self
+            .set
+            .base_token
+            .approve(self.protocol.cozy_router.address(), U256::MAX);
+        let _ = state.execute_evm_tx_and_decode(self.address, router_approve_tx);
+
+        for (_, ptoken) in self.set.ptokens.iter() {
+            let ptoken_approve_tx = ptoken.approve(self.protocol.cozy_router.address(), U256::MAX);
+            let _ = state.execute_evm_tx_and_decode(self.address, ptoken_approve_tx);
+        }
+    }
+
+    fn step(
+        &mut self,
+        state: &State<CozyUpdate, CozyWorld>,
+        channel: &AgentChannelSender<CozyUpdate>,
+    ) {
+        if !self
+            .preferences
+            .reaction_time
+            .time_to_react(state.timestamp(), &mut self.rng)
+        {
+            return;
+        }
```

Comment:
Is there a case where an arbitrageur would want to do nothing when it's time to react? I guess no..

Reply:
They will always check if at least it would make sense to execute a trade -- compute the current cost/refund and compare to their fair price. But they only execute when it the current cost/refund is smaller/larger than their fair price. So, they only trade a small fraction of the time.